### DISCUSSION TITLE
Reflection이 무엇인가요? 어떤 단점이 있는지와 함께 설명해주세요

### DISCUSSION THREAD

**Discussion by jimi567** [2024-woowacourse-study/level-interview#52] [Java]
.

**Comment by alstn113** [comment]
리플렉션(Reflection)은 프로그램이 자신의 구조를 조사하고 수정할 수 있는 능력을 가리킵니다.
즉, 실행 중에 클래스의 정보를 알아내거나 수정할 수 있는 기능을 제공합니다.
리플렉션을 사용하면 프로그램이 런타임 시점에 동적으로 클래스를 조작할 수 있습니다.

주요 사용 사례
1. 런타임 시점(프로그램 실행 중)에 클래스의 이름, 필드, 메서드, 어노테이션등에 대한 정보를 동적으로 얻을 수 있다.
2. 클래스의 생성자를 호출하여 객체를 생성하거나, 메서드를 호출하여 객체의 동작을 실행할 수 있다.
3. 객체의 필드 값을 읽거나 수정할 수 있다.
4. 클래스나 메서드에 적용된 어노테이션 정보를 동적으로 확인하고 처리할 수 있다.

단점
1. 컴파일 시간에는 알 수 없는 정보를 찾고 호출하기 때문에 일반적인 메서드 호출보다 성능이 떨어질 수 있다.
2. private 필드나 메서드에 접근할 수 있기 때문에 보안 문제가 있다.
3. 코드의 가독성과 유지보수성이 저하될 수 있다. 정적 타입 검사를 받지 않기 때문에 컴파일러가 코드를 검증할 수 없다.

**Comment by GIVEN53** [comment]
프로그램 실행 중에 클래스의 정보를 검사하고 조작할 수 있는 기능을 제공하는 Java API이다.
클래스의 구조, 필드, 메소드, 생성자 등을 동적으로 검사하고 실행할 수 있다.
컴파일된 바이트 코드를 통해 런타임에 특정 Class의 정보를 추출하기 때문에 동적 바인딩을 제공한다.

### 장점
- 클래스를 동적으로 로드하고 인스턴스를 생성할 수 있다.
- 실행 중에 객체의 상태를 검사하고 조작할 수 있으므로 디버깅 및 테스팅에 유용하다.

### 단점
- 컴파일 타임에 타입 확인이나 예외 검사를 할 수 없다.
- 객체 내부를 직접 접근, 조작하기 때문에 캡슐화를 깨뜨릴 수 있다.
- 코드의 가독성을 떨어뜨린다.

**Comment by le2sky** [comment]
https://tecoble.techcourse.co.kr/post/2020-07-16-reflection-api/

**Comment by GIVEN53** [comment]
런타임에 클래스 검사, 조작
필드,  메서드, 생성자 등
동적 바인딩
디버깅, 테스팅 용이
컴파일 타임에 예외 불가
캡슐화 깨뜨림
가독성 감소
